clang/lib/Sema/TreeTransform.h

```diff
@@ -7668,6 +7668,39 @@ QualType TreeTransform<Derived>::TransformAttributedType(TypeLocBuilder &TLB,
       }
     }
 
+    /* TO_UPSTREAM(BoundsSafety) ON*/
+    // This is primarily here for the `TemplateInstantiator` subclass
+    // where we want to check that after template instantiation the
+    // bounds-safety attributes are being applied to pointer types.
+    if (SemaRef.LangOpts.isBoundsSafetyAttributeOnlyMode()) {
+      const char *AttrName = nullptr;
+      switch (TL.getAttrKind()) {
+      case attr::PtrSingle:
+        AttrName = "__single";
+        LLVM_FALLTHROUGH;
+      case attr::PtrUnsafeIndexable: {
+        AttrName = AttrName ? AttrName : "__unsafe_indexable";
+        QualType ShouldBePtrTy =
+            modifiedType.getDesugaredType(SemaRef.getASTContext());
+        if (!ShouldBePtrTy->isPointerType()) {
```

Comment:
I believe the idiomatic way to check this is to check whether `modifiedType->getAs<PointerType>()` is nonnull